### PR TITLE
Merge upstream 5.6.40 into Percona Server

### PR DESCRIPTION
**Pull request by inikep**
Merge blueprint:
https://jira.percona.com/browse/PS-312

***
Backport `rpl/t/rpl_bug33931.test` from 5.7 as "SHOW SLAVE STATUS"
reports I/O thread failure after changes with
https://jira.percona.com/browse/PS-3593

***
Merged bug86260.test and sp_debug.test

***
Revert Percona fix at
https://jira.percona.com/projects/PS/issues/PS-3903

***
Updated man pages from MySQL Server 5.6.40 source tarball.

***
Updated 'scripts/fill_help_tables.sql' from MySQL Server 5.6.40 source
tarball.

***
VERSION raised to "5.6.40.83.2".
univ.i version raised to "83.2"

### CODE REVIEW

**Comment by inikep** [comment]
https://jenkins.percona.com/job/percona-server-5.6-param/2174/

**Review by laurynas-biveinis — APPROVED** [review]

**Comment by inikep** [comment]
Updated:
```
	modified:   mysql-test/suite/rpl/r/rpl_bug68490.result
	modified:   mysql-test/suite/rpl/r/rpl_gtid_deployment_step.result
	modified:   mysql-test/suite/rpl/t/rpl_bug68490.test
	modified:   mysql-test/suite/rpl/t/rpl_gtid_deployment_step.test
```

**Comment by inikep** [comment]
https://jenkins.percona.com/job/percona-server-5.6-param/2178/
https://jenkins.percona.com/job/percona-server-5.6-param/2179/

**Review by laurynas-biveinis — APPROVED** [review]

File: sql/sql_acl.cc [hunk 1]
```diff
@@ -1,4 +1,4 @@
-/* Copyright (c) 2000, 2017, Oracle and/or its affiliates. All rights reserved.
```
Comment (percona-ysorokin) [inline sql/sql_acl.cc:1]: A merge conflict in `sql/sql_acl.cc` inside `replace_user_table()` is not resolved properly.
```
  if (acl_is_utility_user(combo->user.str, combo->host.str, NULL))
  {
    my_error(ER_NONEXISTING_GRANT, MYF(0), combo->user.str, combo->host.str);
    goto end;
  }

  if (!table->key_info)
  {
    my_error(ER_TABLE_CORRUPT, MYF(0), table->s->db.str,
             table->s->table_name.str);
    goto end;
  }
```
should become just
```
  if (acl_is_utility_user(combo->user.str, combo->host.str, NULL))
  {
    my_error(ER_NONEXISTING_GRANT, MYF(0), combo->user.str, combo->host.str);
    goto end;
  }
```

File: sql/sp_instr.cc [hunk 2]
```diff
@@ -27,8 +27,8 @@
 #include "sql_parse.h"    // check_table_access
 #include "sql_prepare.h"  // reinit_stmt_before_use
 #include "transaction.h"  // trans_commit_stmt
-#include "sql_audit.h"
 #include "debug_sync.h"   // DEBUG_SYNC
+#include "sql_audit.h"
```
Comment (percona-ysorokin) [inline sql/sp_instr.cc:31]: Any reason why we changed header inclusion order here?

File: mysql-test/suite/rpl/t/rpl_gtid_deployment_step.test [hunk 3]
```diff
@@ -73,11 +73,6 @@ select @@global.gtid_executed;
 -- connection master
 call master_procedure();
 select @@global.gtid_executed;
--- connection slave
--- disable_warnings
--- source include/start_slave.inc
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_gtid_deployment_step.test:78]: This looks suspicious, we basically remove starting slave here which is a significant change in the behavior.
Consider adding
`--let $slave_io_errno= convert_error(XXX)`
or
`--let $slave_sql_errno= convert_error(XXX)`
before `-- source include/start_slave.inc` instead.

File: mysql-test/suite/rpl/t/rpl_gtid_deployment_step.test [hunk 4]
```diff
@@ -103,10 +98,6 @@ select @@global.gtid_executed;
 call master_procedure();
 -- replace_regex /[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}/UUID/
 select @@global.gtid_executed;
--- connection slave
--- disable_warnings
--- source include/start_slave.inc
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_gtid_deployment_step.test:108]: Likewise

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]

File: sql/sp_instr.cc (hunk 2 above)
Comment (inikep) [reply sql/sp_instr.cc:31]: `sql_audit.h` was added by Percona and above change keeps code more similar to upstream.

Comment (percona-ysorokin) [reply sql/sp_instr.cc:31]: OK

File: mysql-test/suite/rpl/t/rpl_gtid_deployment_step.test (hunk 3 above)
Comment (inikep) [reply mysql-test/suite/rpl/t/rpl_gtid_deployment_step.test:78]: In this case slave is already running and master was restarted. `start_slave.inc` can return Last_IO_Errno=2003 that is `error reconnecting to master 'root@127.0.0.1:13001' - retry-time: 1  retries: 6`.
To solve this issue `start_slave.inc` should be replaced with `sync_slave_sql_with_master.inc` that reconnects slave with master, but `sync_slave_sql_with_master.inc` is already used a few lines later.

File: sql/sql_acl.cc (hunk 1 above)
Comment (inikep) [reply sql/sql_acl.cc:1]: fixed

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Comment by inikep** [comment]
https://jenkins.percona.com/job/percona-server-5.6-param/2180/

**Pull request merged**